conda/core/solve.py

```diff
@@ -2153,6 +2155,155 @@ def working_state_reset(self):
         self.final_environment_specs = None
 
 
+class TrackedMap(MutableMapping):
```

Comment:
This should probably be a UserDict to reduce the implementation here a notch.

Reply:
That was the plan originally, but I decided against it to simplify the implementation. If we used `UserDict`, we would still have to reimplement everything to get the correct stack associated with the log, or we would have to have some magic that automatically calculates the correct stack level for the log call. To keep things easily maintainable I opted against the stack level calculation, I think reimplementing the methods is better as it is less complex and easier to read and understand the code. Since we would be reimplementing the methods, while promising a `dict` subclass, I thought it would be better to move to `MutableMapping` and limiting the API surface. There's no specific reason here to promise a `dict`.

Anyway, the `MutableMapping` implementation would also be pretty simple if we didn't have to override the mixin methods.